packages/core/mesh/teleport/src/muxing/balancer.ts

```diff
@@ -140,42 +115,80 @@ export class Balancer {
     return this._channels[index];
   }
 
-  private _getNextChunk(): ChunkEnvelope {
+  private _enqueueChunk(data: Uint8Array, trigger: Trigger, channelId: number) {
+    if (!this._channels.includes(channelId)) {
+      throw new Error(`Unknown channel ${channelId}`);
+    }
+
+    if (!this._sendBuffers.has(channelId)) {
+      this._sendBuffers.set(channelId, []);
+    }
+
+    const sendBuffer = this._sendBuffers.get(channelId)!;
+
+    const chunks = [];
+    for (let idx = 0; idx < data.length; idx += MAX_CHUNK_SIZE) {
+      chunks.push(data.subarray(idx, idx + MAX_CHUNK_SIZE));
+    }
+
+    chunks.forEach((chunk, index) => {
+      const msg = encodeChunk({
+        chunk,
+        channelId,
+        dataLength: index === 0 ? data.length : undefined,
+      });
+      sendBuffer.push({ msg, trigger: index === chunks.length - 1 ? trigger : undefined });
+    });
+  }
+
+  // get the next chunk or null if there are no chunks remaining
+
+  private _getNextChunk(): ChunkEnvelope | null {
     let chunk;
-    while (!chunk) {
+    while (this._sendBuffers.size > 0) {
       const channelId = this._getNextCallerId();
       const sendBuffer = this._sendBuffers.get(channelId);
       if (!sendBuffer) {
         continue;
       }
 
       chunk = sendBuffer.shift();
+      if (!chunk) {
+        continue;
+      }
       if (sendBuffer.length === 0) {
         this._sendBuffers.delete(channelId);
       }
+      return chunk;
     }
-    return chunk;
+    return null;
   }
 
-  private async _sendChunk() {
-    if (this._sendBuffers.size === 0) {
+  private async _sendChunk(channelId: number) {
+    if (this._sending) {
       return;
     }
-
-    if (!this._framer.writable) {
-      await this._framer.drain.waitForCount(1);
-    }
-
-    const chunk = this._getNextChunk();
-
-    try {
-      await this._framer.port.send(chunk.msg);
-      chunk.trigger?.wake();
-    } catch (err: any) {
-      chunk.trigger?.throw(err);
+    this._sending = true;
+    let chunk: ChunkEnvelope | null;
+    chunk = this._getNextChunk();
+    while (chunk) {
+      // this is not needed since we await the chunk send
+      if (!this._framer.writable) {
```

Comment:
👍

Reply:
does this mean we could remove `framer.drain` and `framer.writable` if they are unused?